File: code/apipython/bd.py
import MySQLdb
import MySQLdb.cursors
import json
import os
import collections
import datetime
import constantes

class banco(object):
    __host     = constantes._BD_HOST_
    __db_name  = constantes._BD_DATABASE_
    __username = constantes._BD_LOGIN_
    __password = constantes._BD_PASSWORD_
    __conn     = None
    __stmt     = None
    __tipo     = "QUERY"
    __erro     = ""
    __rowcount = 0
    __lastid   = 0
    def __init__(self):
        self.__tipo = "QUERY"

    def conectar(self):
        self.__erro=""
        try:
            self.__conn = MySQLdb.connect(host=self.__host, user=self.__username, passwd=self.__password, db=self.__db_name)
        except Exception as exception:
            self.__erro = 'Não foi possivel conectar com o Banco de Dados'#$exception->getMessage();

    def desconectar(self):
        if self.__stmt!=None:
            self.__stmt.close()
        self.__stmt = None
        if self.__conn!=None:
            self.__conn.close()   
        self.__conn = None

    def temErro(self):
        return self.__erro!=""

    def getErro(self):
        return self.__erro

    def prepara(self, query, valores=None):
        try:
            self.__rowcount = -1
            self.__lastid   = -1
            self.conectar()
            if (self.temErro()):
                return
            self.__stmt = self.__conn.cursor(cursorclass=MySQLdb.cursors.DictCursor)
            self.__stmt.execute(query, valores)
            query = query.lower()
            
            if query.find("insert")     != -1:
                self.__tipo = 'INSERT'
            elif query.find("update")   != -1:
                self.__tipo = 'UPDATE'
            elif query.find("delete")   != -1:
                self.__tipo = 'DELETE' 
            else:
                self.__tipo = 'QUERY'
        except Exception as exception:
            self.__erro = 'Erro na preparação da consulta ao banco de dados'+str(exception)#$exception->getMessage();
 
    def count(self):
        try:
            if self.temErro():
                return -1
            else:
                return self.__rowcount
        except Exception as exception:
            self.__erro = 'Erro ao tentar verificar a quantidade de elementos da consulta ao banco de dados'+str(exception)
    
    def lastid(self):
        try:
            if self.temErro():
                return -1
            else:
                return self.__lastid
        except Exception as exception:
            self.__erro = 'Erro ao tentar verificar o ultimo id inserido no banco de dados'+str(exception)
                    
    def executar(self):
        try:
            if self.temErro():
                return
            if self.__tipo=="QUERY":
                data = self.__stmt.fetchall() 
                self.__rowcount = self.__stmt.rowcount
            elif self.__tipo=="INSERT":
                self.__rowcount = self.__stmt.rowcount
                data = self.__lastid = self.__stmt.lastrowid
                self.__conn.commit()
            elif self.__tipo=="UPDATE":
                data = self.__rowcount = self.__stmt.rowcount
                self.__conn.commit()
            elif self.__tipo=="DELETE":
                data = self.__rowcount = self.__stmt.rowcount
                self.__conn.commit()
            else:
                raise Exception("deu problema")
            self.__stmt.close()
            self.__stmt = None    
            return data
        except Exception as exception:
            self.__conn.rollback()
            desconectar()
            self.__erro = 'Erro na execução da consulta ao banco de dados'+str(exception)
            
    def listar(self, query):
        try:
            self.prepara(query)
            return self.executar()
        except Exception as exception:
            self.__erro = 'Erro ao tentar listar os elementos do banco de dados'

    def obterPeloId(self, query, id):
        try:
            self.prepara(query, (id))
            vet = self.executar()
            if self.temErro():
                return -1
            return vet[0]
        except Exception as exception:
            self.__erro = 'Erro ao obter um elemento do banco de dados'

File: code/apinode/bd.js

const mysql = require('mysql');
const constantes = require('./constantes');


class banco {
    /*var host     = constantes._BD_HOST_;
    var db_name  = constantes._BD_DATABASE_;
    var username = constantes._BD_LOGIN_;
    var password = constantes._BD_PASSWORD_;
    var conn     = null;
    var stmt     = null;
    var tipo     = "QUERY";
    var erro     = "";
    var rowcount = 0;
    var lastid   = 0;
    var sql      = "";
    var params   = null;*/

    constructor()
    {
        this.host     = constantes._BD_HOST_;
        this.db_name  = constantes._BD_DATABASE_;
        this.username = constantes._BD_LOGIN_;
        this.password = constantes._BD_PASSWORD_;
        this.conn     = null;
        this.stmt     = null;
        this.tipo     = "QUERY";
        this.erro     = "";
        this.rowcount = 0;
        this.lastid   = 0;
        this.sql      = "";
        this.params   = null;
    }

    conectar(){
        try{
            this.conn = mysql.createConnection({ host:this.host, user:this.username, password:this.password, database:this.db_name});
        }
        catch (exception){
            this.erro = 'Não foi possivel conectar com o Banco de Dados' + exception;
        }
    }

    desconectar(){
        try{
            if (this.stmt!=null)
                this.stmt.close();
            this.stmt = null;
            if (this.conn!=null)
                this.conn.end();   
            this.conn = null;
        }
        catch (exception){
            this.erro = 'Não foi possivel conectar com o Banco de Dados' + exception;
        }
    }

    temErro(){
        return this.erro!="";
    }

    getErro(){
        return this.erro;
    }

    prepara(query, values=null){
        try{
            this.rowcount = -1;
            this.lastid   = -1;
            //this.conectar();
            this.sql = query;
            this.params = values;
            if (this.temErro())
                return;
            query = query.toLowerCase();            
            if (query.indexOf("insert")!= -1)
                this.tipo = 'INSERT';
            else if (query.indexOf("update")!= -1)
                this.tipo = 'UPDATE';
            else if (query.indexOf("delete")!= -1)
                this.tipo = 'DELETE'; 
            else
                this.tipo = 'QUERY';
        }catch(exception){
            this.erro = 'Erro na preparação da consulta ao banco de dados'+exception;
        }
    }

    count(){
        try{
            if (this.temErro())
                return -1;
            else
                return this.rowcount;
        }catch(exception){
            this.erro = 'Erro ao tentar verificar a quantidade de elementos da consulta ao banco de dados'+exception;
        }
    }

    lastid(){
        try{
            if (this.temErro())
                return -1;
            else
                return this.lastid;
        }catch(exception){
            this.erro = 'Erro ao tentar verificar o ultimo id inserido no banco de dados'+exception;
        }
    }
                    
    executar(async){
        try{
            var self = this;
            if (this.temErro())
                async(null);
            this.conn.query(this.sql, this.params, function(err, result){
                self.rowcount = -1;
                self.lastid = -1;
                if (err) 
                    self.erro = err;
                else if (self.tipo=="QUERY")
                    self.rowcount = result.length;
                else if (self.tipo=="INSERT"){
                    self.rowcount = result.affectedRows;
                    self.lastid = result.insertId;
                }else if (self.tipo=="UPDATE")
                    self.rowcount = result.affectedRows;
                else if (self.tipo=="DELETE")
                    self.rowcount = result.affectedRows;
                else
                    self.erro = "deu problema";
                async(result);
            });
        }catch(exception){
            this.desconectar();
            this.erro = 'Erro na execução da consulta ao banco de dados'+exception;
            async(null);
        }
    }
            
    listar(query, async){
        try{
            this.prepara(query);
            this.executar(async);
        }catch(exception){
            this.erro = 'Erro ao tentar listar os elementos do banco de dados';
            async(null);
        }
    }

    obterPeloId(query, id, async){
        try{
            this.prepara(query, [id]);
            this.executar(async);
        }catch(exception){
            this.erro = 'Erro ao obter um elemento do banco de dados';
            async(null);
        }
    }
}

module.exports = banco;